File: 343/343_version1.py
class Solution:
    def integerBreak(self, n: int) -> int:
        res = [0 for _ in range(n+1)]
        res[1] = 1
        res[2] = 1
        index = 3
        while index <= n:
            temp = 2
            while temp <= index//2+1:
                curr1 = temp*(index-temp)
                curr2 = temp*res[index-temp]
                res[index] = max([res[index],curr1,curr2])
                temp+=1
            index+=1
        return res[n]

File: 343/343_version1.java
class Solution {
    public int integerBreak(int n) {
        int[] cache = new int[n+1];
        cache[1] = 1;cache[2] = 1;
        int curr = 3;
        while(curr <= n){
            int res = 0;
            for(int temp = 1;temp <= curr/2+1;temp++){
                int option1 = temp * (curr-temp);
                int option2 = temp * cache[curr-temp];
                res = Math.max(res,Math.max(option1,option2));
            }
            cache[curr] = res;
            curr++;
        }
        return cache[n];
    }
}
